File: splendor/reference/python/validate_data.py
"""
Code-level validator for splendor_data.json.

The JSON Schema (../splendor_data.schema.json) is structural only. This script
performs the checks JSON Schema CANNOT express and is the authoritative data
gate for CI. Run BOTH the schema and this file.

Run:  python3 validate_data.py
Exits 0 if all checks pass, 1 otherwise.
"""
import json, os, sys
from collections import Counter, defaultdict

COLORS = ["white", "blue", "green", "red", "black"]
HERE = os.path.dirname(os.path.abspath(__file__))
DATA_PATH = os.path.join(HERE, "..", "splendor_data.json")
SCHEMA_PATH = os.path.join(HERE, "..", "splendor_data.schema.json")

failures = []
def check(cond, msg):
    print(("  PASS " if cond else "  FAIL ") + msg)
    if not cond:
        failures.append(msg)


def main():
    data = json.load(open(DATA_PATH))
    cards = data["cards"]
    nobles = data["nobles"]

    # Optional: run the JSON Schema too, if jsonschema is installed.
    try:
        import jsonschema
        jsonschema.validate(data, json.load(open(SCHEMA_PATH)))
        print("  PASS JSON Schema (structural) validates")
    except ImportError:
        print("  SKIP jsonschema not installed (pip install jsonschema) — structural check skipped")
    except Exception as e:
        check(False, f"JSON Schema validation: {e}")

    print("\n[cards] structure")
    check(len(cards) == 90, "exactly 90 cards")
    ids = [c["id"] for c in cards]
    check(sorted(ids) == list(range(1, 91)), "card ids are exactly 1..90 (unique, contiguous)")
    per_tier = Counter(c["tier"] for c in cards)
    check(per_tier.get(1) == 40 and per_tier.get(2) == 30 and per_tier.get(3) == 20,
          "tier split is 40 / 30 / 20")
    ptc = defaultdict(Counter)
    for c in cards:
        ptc[c["tier"]][c["bonus"]] += 1
    check(all(ptc[1][c] == 8 for c in COLORS), "tier 1: 8 cards per color")
    check(all(ptc[2][c] == 6 for c in COLORS), "tier 2: 6 cards per color")
    check(all(ptc[3][c] == 4 for c in COLORS), "tier 3: 4 cards per color")

    print("\n[cards] points & costs")
    dist = defaultdict(Counter)
    for c in cards:
        dist[c["tier"]][c["points"]] += 1
    check(dict(dist[1]) == {0: 35, 1: 5}, "tier 1 point distribution {0:35, 1:5}")
    check(dict(dist[2]) == {1: 10, 2: 15, 3: 5}, "tier 2 point distribution {1:10, 2:15, 3:5}")
    check(dict(dist[3]) == {3: 5, 4: 10, 5: 5}, "tier 3 point distribution {3:5, 4:10, 5:5}")
    check(sum(c["points"] for c in cards) == 140, "total prestige across all cards == 140")
    check(all(set(c["cost"].keys()) == set(COLORS) for c in cards), "every cost has exactly the 5 gem colors (no gold)")
    check(all(0 <= v <= 7 for c in cards for v in c["cost"].values()), "all cost values in 0..7")
    check(all(c["bonus"] in COLORS for c in cards), "all bonuses are valid gem colors")
    # official rulebook example card: Tier-3 blue, 4 pts, 6 white + 3 blue + 3 black
    example = [c for c in cards if c["tier"] == 3 and c["bonus"] == "blue" and c["points"] == 4
               and c["cost"] == {"white": 6, "blue": 3, "green": 0, "red": 0, "black": 3}]
    check(len(example) == 1, "official rulebook example card present (T3 blue, 4pt, 6w+3u+3k)")
    # tier-3 average = 4.0 (published)
    t3 = [c for c in cards if c["tier"] == 3]
    check(abs(sum(c["points"] for c in t3) / len(t3) - 4.0) < 1e-9, "tier-3 average points == 4.0")

    print("\n[nobles]")
    check(len(nobles) == 10, "exactly 10 nobles")
    check(sorted(n["id"] for n in nobles) == list(range(1, 11)), "noble ids are exactly 1..10")
    check(all(n["points"] == 3 for n in nobles), "every noble worth 3 points")
    shape_ok = True
    for n in nobles:
        vals = [n["requirement"][c] for c in COLORS]
        nonzero = [v for v in vals if v > 0]
        is_4_4 = (sorted(nonzero) == [4, 4])
        is_3_3_3 = (sorted(nonzero) == [3, 3, 3])
        if not (is_4_4 or is_3_3_3):
            shape_ok = False
    check(shape_ok, "every noble is exactly two colors x4 OR three colors x3")
    appearances = {c: sum(1 for n in nobles if n["requirement"][c] > 0) for c in COLORS}
    check(appearances == {c: 5 for c in COLORS}, f"noble color appearances symmetric (5 each); got {appearances}")

    print("\n[meta.checksums cross-check]")
    cs = data.get("meta", {}).get("checksums", {})
    if cs:
        check(cs.get("totalCards") == 90, "meta.checksums.totalCards == 90")
        check(cs.get("totalPrestigeOnAllCards") == 140, "meta.checksums.totalPrestige == 140")
        check(cs.get("nobleColorAppearances") == {c: 5 for c in COLORS}, "meta.checksums noble appearances == 5 each")
    else:
        print("  SKIP no meta.checksums present")

    print()
    if failures:
        print(f"RESULT: FAIL ({len(failures)} check(s) failed)")
        sys.exit(1)
    print("RESULT: PASS — all code-level data checks passed")
    sys.exit(0)


if __name__ == "__main__":
    main()

File: splendor/packages/engine/src/validateData.ts
// Code-level data validation (SPEC §11/§19) — the checks JSON Schema cannot
// express. Returns a list of failure messages (empty array => valid).

import { Card, Color, COLORS, GameData, Noble, Tier } from "./types";

const TIERS: Tier[] = [1, 2, 3];

export function validateGameData(data: GameData): string[] {
  const fail: string[] = [];
  const check = (cond: boolean, msg: string) => {
    if (!cond) fail.push(msg);
  };
  const cards: Card[] = data.cards;
  const nobles: Noble[] = data.nobles;

  // cards: structure
  check(cards.length === 90, "exactly 90 cards");
  const ids = cards.map((c) => c.id).sort((a, b) => a - b);
  check(JSON.stringify(ids) === JSON.stringify(Array.from({ length: 90 }, (_, i) => i + 1)),
    "card ids are exactly 1..90 (unique, contiguous)");
  const perTier: Record<number, number> = {};
  for (const c of cards) perTier[c.tier] = (perTier[c.tier] ?? 0) + 1;
  check(perTier[1] === 40 && perTier[2] === 30 && perTier[3] === 20, "tier split is 40 / 30 / 20");
  for (const t of TIERS) {
    for (const col of COLORS) {
      const n = cards.filter((c) => c.tier === t && c.bonus === col).length;
      const expect = t === 1 ? 8 : t === 2 ? 6 : 4;
      check(n === expect, `tier ${t}: ${expect} cards of ${col} (got ${n})`);
    }
  }

  // cards: points & costs
  const dist = (t: Tier) => {
    const m: Record<number, number> = {};
    for (const c of cards) if (c.tier === t) m[c.points] = (m[c.points] ?? 0) + 1;
    return m;
  };
  check(JSON.stringify(dist(1)) === JSON.stringify({ 0: 35, 1: 5 }), "tier 1 point distribution {0:35,1:5}");
  check(JSON.stringify(dist(2)) === JSON.stringify({ 1: 10, 2: 15, 3: 5 }), "tier 2 point distribution {1:10,2:15,3:5}");
  check(JSON.stringify(dist(3)) === JSON.stringify({ 3: 5, 4: 10, 5: 5 }), "tier 3 point distribution {3:5,4:10,5:5}");
  check(cards.reduce((s, c) => s + c.points, 0) === 140, "total prestige across all cards == 140");
  check(cards.every((c) => COLORS.every((col) => typeof c.cost[col] === "number")), "every cost has the 5 gem colors");
  check(cards.every((c) => COLORS.every((col) => c.cost[col] >= 0 && c.cost[col] <= 7)), "all cost values in 0..7");
  check(cards.every((c) => (COLORS as readonly string[]).includes(c.bonus)), "all bonuses are valid gem colors");
  const example = cards.filter(
    (c) => c.tier === 3 && c.bonus === "blue" && c.points === 4 &&
      c.cost.white === 6 && c.cost.blue === 3 && c.cost.black === 3 && c.cost.green === 0 && c.cost.red === 0,
  );
  check(example.length === 1, "official rulebook example card present (T3 blue, 4pt, 6w+3u+3k)");
  const t3 = cards.filter((c) => c.tier === 3);
  check(Math.abs(t3.reduce((s, c) => s + c.points, 0) / t3.length - 4.0) < 1e-9, "tier-3 average points == 4.0");

  // nobles
  check(nobles.length === 10, "exactly 10 nobles");
  const nids = nobles.map((n) => n.id).sort((a, b) => a - b);
  check(JSON.stringify(nids) === JSON.stringify(Array.from({ length: 10 }, (_, i) => i + 1)), "noble ids are exactly 1..10");
  check(nobles.every((n) => n.points === 3), "every noble worth 3 points");
  let shapeOk = true;
  for (const n of nobles) {
    const nonzero = COLORS.map((c) => n.requirement[c]).filter((v) => v > 0).sort((a, b) => a - b);
    const is44 = JSON.stringify(nonzero) === JSON.stringify([4, 4]);
    const is333 = JSON.stringify(nonzero) === JSON.stringify([3, 3, 3]);
    if (!is44 && !is333) shapeOk = false;
  }
  check(shapeOk, "every noble is exactly two colors x4 OR three colors x3");
  const appearances: Record<Color, number> = { white: 0, blue: 0, green: 0, red: 0, black: 0 };
  for (const n of nobles) for (const c of COLORS) if (n.requirement[c] > 0) appearances[c] += 1;
  check(COLORS.every((c) => appearances[c] === 5), `noble color appearances symmetric (5 each); got ${JSON.stringify(appearances)}`);

  // meta.checksums cross-check (if present)
  const meta = data.meta as { checksums?: Record<string, unknown> } | undefined;
  const cs = meta?.checksums;
  if (cs) {
    check(cs.totalCards === 90, "meta.checksums.totalCards == 90");
    check(cs.totalPrestigeOnAllCards === 140, "meta.checksums.totalPrestige == 140");
    check(JSON.stringify(cs.nobleColorAppearances) === JSON.stringify({ white: 5, blue: 5, green: 5, red: 5, black: 5 }),
      "meta.checksums noble appearances == 5 each");
  }

  return fail;
}
